File: app/src/main/java/awais/instagrabber/utils/LocaleUtils.java
package awais.instagrabber.utils;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.view.ContextThemeWrapper;

import androidx.annotation.Nullable;

import java.util.Locale;

// taken from my app TESV Console Codes
public final class LocaleUtils {
    private static Locale defaultLocale, currentLocale;

    public static void setLocale(Context baseContext) {
        if (defaultLocale == null) defaultLocale = Locale.getDefault();

        if (baseContext instanceof ContextThemeWrapper)
            baseContext = ((ContextThemeWrapper) baseContext).getBaseContext();

        final String lang = LocaleUtils.getCorrespondingLanguageCode(baseContext);

        currentLocale = TextUtils.isEmpty(lang) ? defaultLocale : new Locale(lang);
        Locale.setDefault(currentLocale);

        final Resources res = baseContext.getResources();
        final Configuration config = res.getConfiguration();

        config.locale = currentLocale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            config.setLocale(currentLocale);
            config.setLayoutDirection(currentLocale);
        }

        res.updateConfiguration(config, res.getDisplayMetrics());
    }

    public static Locale getCurrentLocale() {
        return currentLocale;
    }

    public static void updateConfig(final ContextThemeWrapper wrapper) {
        if (currentLocale != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            final Configuration configuration = new Configuration();
            configuration.locale = currentLocale;
            configuration.setLocale(currentLocale);
            wrapper.applyOverrideConfiguration(configuration);
        }
    }

    @Nullable
    private static String getCorrespondingLanguageCode(final Context baseContext) {
        if (Utils.settingsHelper == null)
            Utils.settingsHelper = new SettingsHelper(baseContext);

        final String appLanguageSettings = Utils.settingsHelper.getString(Constants.APP_LANGUAGE);
        if (TextUtils.isEmpty(appLanguageSettings)) return null;

        final int appLanguageIndex = Integer.parseInt(appLanguageSettings);
        if (appLanguageIndex == 1) return "en";
        if (appLanguageIndex == 2) return "fr";
        if (appLanguageIndex == 3) return "es";
        if (appLanguageIndex == 4) return "zh";
        if (appLanguageIndex == 5) return "in";
        if (appLanguageIndex == 6) return "it";
        if (appLanguageIndex == 7) return "de";
        if (appLanguageIndex == 8) return "pl";
        if (appLanguageIndex == 9) return "tr";
        if (appLanguageIndex == 10) return "pt";
        if (appLanguageIndex == 11) return "fa";
        if (appLanguageIndex == 12) return "mk";
        if (appLanguageIndex == 13) return "vi";

        return null;
    }
}

File: app/src/main/java/awais/instagrabber/utils/LocaleUtils.kt
package awais.instagrabber.utils

import android.content.Context
import android.content.res.Configuration
import android.view.ContextThemeWrapper
import awais.instagrabber.fragments.settings.PreferenceKeys
import java.util.*

// taken from my app TESV Console Codes
object LocaleUtils {
    private lateinit var defaultLocale: Locale

    @JvmStatic
    lateinit var currentLocale: Locale
        private set

    @JvmStatic
    fun setLocale(baseContext: Context) {
        var baseContext1 = baseContext
        defaultLocale = Locale.getDefault()
        if (baseContext1 is ContextThemeWrapper) baseContext1 = baseContext1.baseContext
        if (Utils.settingsHelper == null) Utils.settingsHelper = SettingsHelper(baseContext1)
        val appLanguageSettings = Utils.settingsHelper.getString(PreferenceKeys.APP_LANGUAGE)
        val lang = getCorrespondingLanguageCode(appLanguageSettings)
        currentLocale = when {
            TextUtils.isEmpty(lang) -> defaultLocale
            lang!!.contains("_") -> {
                val split = lang.split("_")
                Locale(split[0], split[1])
            }
            else -> Locale(lang)
        }
        currentLocale.let {
            Locale.setDefault(it)
            val res = baseContext1.resources
            val config = res.configuration
            // config.locale = currentLocale
            config.setLocale(it)
            config.setLayoutDirection(it)
            res.updateConfiguration(config, res.displayMetrics)
        }
    }

    @JvmStatic
    fun updateConfig(wrapper: ContextThemeWrapper) {
        if (!this::currentLocale.isInitialized) return
        val configuration = Configuration()
        // configuration.locale = currentLocale
        configuration.setLocale(currentLocale)
        wrapper.applyOverrideConfiguration(configuration)
    }

    fun getCorrespondingLanguageCode(appLanguageSettings: String): String? {
        if (TextUtils.isEmpty(appLanguageSettings)) return null
        when (appLanguageSettings.toInt()) {
            1 -> return "en"
            2 -> return "fr"
            3 -> return "es"
            4 -> return "zh_CN"
            5 -> return "in"
            6 -> return "it"
            7 -> return "de"
            8 -> return "pl"
            9 -> return "tr"
            10 -> return "pt"
            11 -> return "fa"
            12 -> return "mk"
            13 -> return "vi"
            14 -> return "zh_TW"
            15 -> return "ca"
            16 -> return "ru"
            17 -> return "hi"
            18 -> return "nl"
            19 -> return "sk"
            20 -> return "ja"
            21 -> return "el"
            22 -> return "eu"
            23 -> return "sv"
            24 -> return "ko"
            25 -> return "ar"
        }
        return null
    }
}